### PR TITLE
gRPC: Config grpc memory quota according to the total memory

### PR DESCRIPTION
**Pull request by hunterlxt**
Signed-off-by: Xintao <hunterlxt@live.com>

### What problem does this PR solve?

When the network becomes poor, gRPC can easily cause OOM due to holding too much memory. The default grpc value should be not too large.

### What is changed and how it works?

Use total machine memory to calculate the memory for grpc.

### Release note 
- Auto config grpc memory quota according to the total memory.

### CODE REVIEW

**Comment by BusyJay** [comment]
How to know it's really hold by grpc?

**Comment by hunterlxt** [comment]
> How to know it's really hold by grpc?
It's not that OOM must be held by grpc, the PR is only for this scenario: sent too much data into grpc sink but not send out actually.

**Comment by hunterlxt** [comment]
I found lots of CI failure but not related to this PR.

**Comment by BusyJay** [comment]
> sent too much data into grpc sink but not send out actually

How to verify it's a real/common problem instead of some dragon we imaging? And what value is appropriate? How much memory does it need generally?  Limiting memory by default can cause connection reset if memory usage burst.

**Comment by hunterlxt** [comment]
> > sent too much data into grpc sink but not send out actually
> 
> How to verify it's a real/common problem instead of some dragon we imaging? And what value is appropriate? How much memory does it need generally? Limiting memory by default can cause connection reset if memory usage burst.

The memory really reaches the quota, it may indeed lead to OOM, if not, the quota will not affect anything.

**Comment by BusyJay** [comment]
> The memory really reaches the quota, it may indeed lead to OOM

Why? What's the general memory usage by grpc? Why reaching 10%, which is the default value according to the PR, can lead to OOM *indeed*? What happen to the 90% left?

File: src/server/config.rs [hunk 1]
```diff
@@ -19,7 +19,7 @@ const DEFAULT_STATUS_ADDR: &str = "127.0.0.1:20180";
 const DEFAULT_GRPC_CONCURRENCY: usize = 4;
 const DEFAULT_GRPC_CONCURRENT_STREAM: i32 = 1024;
 const DEFAULT_GRPC_RAFT_CONN_NUM: usize = 1;
-const DEFAULT_GRPC_MEMORY_POOL_QUOTA: u64 = isize::MAX as u64;
+const DEFAULT_GRPC_MEMORY_POOL_QUOTA_RATIO: f64 = 0.1;
```
Comment (Little-Wallace) [inline src/server/config.rs:22]: Is this too large?

Comment (hicqu) [reply src/server/config.rs:22]: I think `cmp::max(total * 0.1, 2GB)` is ok.

**Comment by hicqu** [comment]
Maybe this PR can't prevent all OOMs, but I think it's helpful.

**Comment by hunterlxt** [comment]
Maybe it's better to quota than nothing to do (unlimited).

**Comment by BusyJay** [comment]
> What value is appropriate? How much memory does it need generally? 
What's the general memory usage by grpc? 

So can anyone answer the questions above?

**Comment by hunterlxt** [comment]
> > What value is appropriate? How much memory does it need generally?
> > What's the general memory usage by grpc?
> 
> So can anyone answer the questions above?

only answer the 3rd: most of the memory in grpc is used to buffer when read data from tcp.

**Comment by hunterlxt** [comment]
Any progress?

**Comment by hunterlxt** [comment]
There is no discussion progress for the time being, close this PR first.

**Pull request closed without merging**